gr-blocks/lib/rotator_cc_impl.cc

```diff
@@ -40,18 +46,91 @@ void rotator_cc_impl::set_phase_inc(double phase_inc)
     d_r.set_phase_incr(exp(gr_complex(0, phase_inc)));
 }
 
+void rotator_cc_impl::handle_ctrl_msg(pmt::pmt_t msg)
+{
+    gr::thread::scoped_lock l(d_mutex);
+
+    static const pmt::pmt_t inc_key = pmt::intern("inc");
+    static const pmt::pmt_t offset_key = pmt::intern("offset");
+
+    if (!pmt::is_dict(msg)) {
+        throw std::runtime_error("rotator_cc: Rotator control message "
+                                 "must be a PMT dictionary");
+    }
+
+    if (!pmt::dict_has_key(msg, inc_key)) {
+        throw std::runtime_error("rotator_cc: Phase increment control "
+                                 "message must contain an \"inc\" key");
+    }
+
+    /* Define the new phase increment (mandatory) and the absolute sample offset
+     * when the update is to be applied (optional). If the latter is not defined
+     * in the control message, update the phase increment right away. */
+    phase_inc_update_t update{};
+    update.phase_inc = pmt::to_double(pmt::dict_ref(msg, inc_key, pmt::PMT_NIL));
+    update.offset = pmt::dict_has_key(msg, offset_key)
+                        ? pmt::to_uint64(pmt::dict_ref(msg, offset_key, pmt::PMT_NIL))
+                        : nitems_written(0);
+
+    d_inc_update_queue.push(update);
+}
+
```

Comment:
In the previous implementation, the message port was named `phase_inc` and this method was called `handle_phase_inc_msg`. I've decided to make it more generic to facilitate future expansion. So I named the port `ctrl` and the message handler `handle_ctrl_msg`.

Not sure if this is common practice in the GR codebase, but the idea would be that we could send multiple commands to the rotator. This implementation adds a single command (to change the phase increment), but it would be easy to add other commands (e.g., to set the phase) using the same port and message handler function.

Now, two questions:
- Should we stick to this or should I revert back to a dedicated `phase_inc` port?
- If we use the suggested generic `ctrl` port, should I add a `command` field on the PMT message and call it `set_phase_inc`? The rationale is that, if other commands are added in the future, it would be easier to coordinate them with `command` field on the input `msg`.